File: spec/schemas/cfc/cfc-v_2026-06-16_20-02_00_49b99195/implementation.js
/**
 * CFC v1 Reference Implementation (JavaScript — normative)
 * CFC version: cfc-v_2026-06-16_20-02_00_49b99195
 * Spec:        spec/schemas/cfc/cfc-v_2026-06-16_20-02_00_49b99195/spec-blob.txt
 *
 * This file is the normative JS implementation for CFC Algorithm v1.
 * It is co-packaged in the schema version folder so that any interpreter
 * or compiler can reproduce the exact algorithm used when a document was
 * created, without defaulting to a current (potentially incompatible) version.
 */

async function cidShort(bytes, fileType = "binary") {
  if (fileType === "text") {
    const s = new TextDecoder("utf-8").decode(bytes)
      .replace(/\r\n/g, "\n").replace(/\r/g, "\n");
    bytes = new TextEncoder().encode(s);
  }
  const digest = await crypto.subtle.digest("SHA-256", bytes);
  return Array.from(new Uint8Array(digest))
    .map(b => b.toString(16).padStart(2, "0"))
    .join("")
    .slice(0, 8);
}

async function canonicalFilename(title, bytes, ext, fileType = "binary", ts = new Date()) {
  const slug = title.replace(/[^a-zA-Z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "").toLowerCase();
  const pad = n => String(n).padStart(2, "0");
  const stamp = `${ts.getUTCFullYear()}-${pad(ts.getUTCMonth()+1)}-${pad(ts.getUTCDate())}_${pad(ts.getUTCHours())}-${pad(ts.getUTCMinutes())}_${pad(ts.getUTCSeconds())}`;
  const cid = await cidShort(bytes, fileType);
  return `${slug}_${stamp}_${cid}.${ext}`;
}

function parseCanonicalFilename(filename) {
  const pattern = /^([a-z0-9][a-z0-9_-]*)_([0-9]{4}-[0-9]{2}-[0-9]{2})_([0-9]{2}-[0-9]{2})_([0-9]{2})_([0-9a-f]{8})(?:\.([a-zA-Z0-9]+))?$/;
  const m = filename.match(pattern);
  if (!m) return null;
  return {
    title: m[1],
    timestamp: `${m[2]}_${m[3]}_${m[4]}`,
    cid_short: m[5],
    extension: m[6] ?? null,
  };
}

async function verifyCanonicalFilename(filename, bytes, fileType = "binary") {
  const parsed = parseCanonicalFilename(filename);
  if (!parsed) return false;
  const actual = await cidShort(bytes, fileType);
  return parsed.cid_short === actual;
}

if (typeof module !== "undefined") module.exports = { cidShort, canonicalFilename, parseCanonicalFilename, verifyCanonicalFilename };

File: spec/schemas/cfc/cfc-v_2026-06-16_20-02_00_49b99195/implementation.py
"""
CFC v1 Reference Implementation (Python)
CFC version: cfc-v_2026-06-16_20-02_00_49b99195
Spec:        spec/schemas/cfc/cfc-v_2026-06-16_20-02_00_49b99195/spec-blob.txt

This file is the normative Python implementation for CFC Algorithm v1.
Co-packaged in the schema version folder for schema-locked rendering.
"""

import hashlib, re
from datetime import datetime, timezone


def cid_short(data: bytes, file_type: str = "binary") -> str:
    if file_type == "text":
        text = data.decode("utf-8").replace("\r\n", "\n").replace("\r", "\n")
        data = text.encode("utf-8")
    return hashlib.sha256(data).hexdigest()[:8]


def canonical_filename(title: str, data: bytes, ext: str,
                       file_type: str = "binary",
                       ts: datetime | None = None) -> str:
    if ts is None:
        ts = datetime.now(timezone.utc)
    slug = re.sub(r"[^a-zA-Z0-9]+", "-", title).strip("-").lower()
    stamp = ts.strftime("%Y-%m-%d_%H-%M_%S")
    cid = cid_short(data, file_type)
    return f"{slug}_{stamp}_{cid}.{ext}"


def parse_canonical_filename(filename: str) -> dict | None:
    m = re.match(
        r"^([a-z0-9][a-z0-9_-]*)_(\d{4}-\d{2}-\d{2})_(\d{2}-\d{2})_(\d{2})_([0-9a-f]{8})(?:\.([a-zA-Z0-9]+))?$",
        filename
    )
    if not m:
        return None
    return {
        "title": m.group(1),
        "timestamp": f"{m.group(2)}_{m.group(3)}_{m.group(4)}",
        "cid_short": m.group(5),
        "extension": m.group(6),
    }


def verify_canonical_filename(filename: str, data: bytes,
                               file_type: str = "binary") -> bool:
    parsed = parse_canonical_filename(filename)
    if not parsed:
        return False
    return cid_short(data, file_type) == parsed["cid_short"]
